opnode/rollup/driver/state.go

```diff
@@ -161,98 +167,81 @@ func (s *state) handleNewL1Block(ctx context.Context, newL1Head eth.L1BlockRef)
-func (s *state) createNewL2Block(ctx context.Context) (eth.L1BlockRef, error) {
-	nextOrigin, maxL2Time, err := s.findNextL1Origin(ctx)
+// createNewL2Block builds a L2 block on top of the L2 Head (unsafe). Used by Sequencer nodes to
+// construct new L2 blocks. Verifier nodes will use handleEpoch instead.
+func (s *state) createNewL2Block(ctx context.Context) error {
+	// Figure out which L1 origin block we're going to be building on top of.
+	l1Origin, err := s.findL1Origin(ctx)
 	if err != nil {
 		s.log.Error("Error finding next L1 Origin", "err", err)
-		return eth.L1BlockRef{}, err
+		return err
+	}
+
+	// Rollup is configured to not start producing blocks until a specific L1 block has been
+	// reached. Don't produce any blocks until we're at that genesis block.
+	if l1Origin.Number <= s.Config.Genesis.L1.Number {
+		s.log.Info("Skipping block production because the next L1 Origin is behind the L1 genesis")
+		return nil
 	}
+
+	// Should never happen. Sequencer will halt if we get into this situation somehow.
 	nextL2Time := s.l2Head.Time + s.Config.BlockTime
-	// If we are behind the L1 origin that we should be using then we broke the invariant
-	if nextL2Time < nextOrigin.Time {
+	if nextL2Time < l1Origin.Time {
 		s.log.Error("Cannot build L2 block for time before L1 origin",
-			"l2Head", s.l2Head, "nextL2Time", nextL2Time, "l1Origin", nextOrigin, "l1OriginTime", nextOrigin.Time)
-		return eth.L1BlockRef{}, fmt.Errorf("cannot build L2 block on top %s for time %d before L1 origin %s at time %d",
-			s.l2Head, nextL2Time, nextOrigin, nextOrigin.Time)
+			"l2Head", s.l2Head, "nextL2Time", nextL2Time, "l1Origin", l1Origin, "l1OriginTime", l1Origin.Time)
+		return fmt.Errorf("cannot build L2 block on top %s for time %d before L1 origin %s at time %d",
+			s.l2Head, nextL2Time, l1Origin, l1Origin.Time)
 	}
 
-	// If the L1 origin changed this block, then we are in the first block of the epoch
-	isFirstEpochBlock := s.l2Head.L1Origin.Number != nextOrigin.Number
-
-	// We create at least 1 block per epoch. After that we have to enforce the max timestamp.
-	if !isFirstEpochBlock && nextL2Time >= maxL2Time {
-		s.log.Warn("Skipping block production because we have no slack left to sequence more blocks on the L1 origin",
-			"l2Head", s.l2Head, "nextL2Time", nextL2Time, "l1Origin", nextOrigin, "l1OriginTime", nextOrigin.Time)
-		return eth.L1BlockRef{}, nil
-	}
-	// Don't produce blocks until past the rollup-genesis block of the L1 chain
-	if nextOrigin.Number <= s.Config.Genesis.L1.Number {
-		s.log.Info("Skipping block production because the next L1 Origin is behind the L1 genesis")
-		return eth.L1BlockRef{}, nil
-	}
-	// Actually create the new block
-	newUnsafeL2Head, batch, err := s.output.createNewBlock(ctx, s.l2Head, s.l2SafeHead.ID(), s.l2Finalized, nextOrigin)
+	// Actually create the new block.
+	newUnsafeL2Head, batch, err := s.output.createNewBlock(ctx, s.l2Head, s.l2SafeHead.ID(), s.l2Finalized, l1Origin)
```

Comment:
I think this changes loses the max timestamp enforcement (aka "sequencer drift" in the spec). Unless this is handled somewhere else?

A scenario where this could happen: L1 stops updating, but we keep creating L2 blocks.